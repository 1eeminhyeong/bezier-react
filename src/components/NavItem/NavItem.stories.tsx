/* External dependencies */
import React, { useMemo } from 'react'
import base from 'paths.macro'
import { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { getTitle } from 'Utils/storyUtils'
import { Icon, IconSize } from 'Components/Icon'
import { TagBadgeSize, TagBadgeVariant, Badge } from 'Components/TagBadge'
import NavItem from './NavItem'
import type NavItemProps from './NavItem.types'

export default {
  title: getTitle(base),
  component: NavItem,
} as Meta

const Template: Story<NavItemProps> = (args) => {
  const badge = useMemo(() => (
    <Badge
      size={TagBadgeSize.M}
      variant={TagBadgeVariant.Green}
    >
      Trial
    </Badge>
  ), [])

  return (
    <div style={{ width: '240px', padding: '6px' }}>
      <NavItem {...args} key="general">
        <NavItem content="채널 프로필" />
        <NavItem content="버튼 설치 및 설정" />
        <NavItem content="고객 프로필 데이터" />
      </NavItem>
      <NavItem
        key="member"
        open
        name="member"
        content="채널 프로필"
        leftIcon="group-filled"
      >
        <NavItem content="멤버" />
        <NavItem content="팀" />
        <NavItem content="역할" />
        <NavItem content="봇" />
      </NavItem>
      <NavItem
        key="service"
        open={false}
        name="service"
        content="서비스 구독 및 결제"
        leftIcon="coin"
        rightContent={badge}
      >
        <NavItem content="멤버" />
      </NavItem>
      <NavItem
        key="chat"
        open={false}
        name="chat"
        content="상담"
        leftIcon="chat-progress"
      >
        <NavItem content="멤버" />
      </NavItem>
      <NavItem
        key="integration"
        open={false}
        name="integration"
        content="연동"
        leftIcon="transfer"
      >
        <NavItem content="멤버" />
      </NavItem>
      <NavItem
        key="resource"
        open={false}
        name="resource"
        content="리소스"
        leftIcon="book"
      >
        <NavItem content="멤버" />
      </NavItem>
      <NavItem
        key="security"
        open={false}
        name="security"
        content="보안 및 개발"
        leftIcon="tool"
      >
        <NavItem content="멤버" />
      </NavItem>
    </div>
  )
}

export const Primary: Story<NavItemProps> = Template.bind({})

Primary.args = {
  open: true,
  active: false,
  name: 'general',
  content: '일반 설정',
  leftIcon: 'settings',
  leftIconColor: 'txt-black-dark',
  rightContent: <Icon name="error-filled" size={IconSize.XS} color="bgtxt-orange-normal" />,
}
