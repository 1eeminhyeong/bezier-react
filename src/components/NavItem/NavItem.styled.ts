/* Internal dependencies */
import { styled, css, ellipsis } from 'Foundation'
import type { InterpolationProps } from 'Types/Foundation'
import { Text } from 'Components/Text'

interface WrapperProps extends InterpolationProps {
  active: boolean
  open: boolean
  paddingLeft: number
}

export const LeftIconWrapper = styled.div`
  display: flex;
  align-items: center;
  margin-right: 10px;
`

export const ContentWrapper = styled(Text)`
  color: ${({ foundation }) => foundation?.theme?.['txt-black-darkest']};
  ${ellipsis()}
`

export const ChevronWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: 2px;
`

export const RightContentWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: auto;
`

const activeItemStyle = css`
  color: ${({ foundation }) => foundation?.theme?.['bgtxt-blue-normal']};
  background-color: ${({ foundation }) => foundation?.theme?.['bgtxt-blue-lightest']};
`

const nonActiveItemStyle = css`
  &:hover {
    background-color: ${({ foundation }) => foundation?.theme?.['bg-black-lighter']};
  }
`

const closedItemStyle = css`
  ${ChevronWrapper} {
    visibility: hidden;
  }

  &:hover {
    ${ChevronWrapper} {
      visibility: visible;
    }
  }
`

export const ItemWrapper = styled.div<WrapperProps>`
  display: flex;
  align-items: center;
  height: 28px;
  padding-right: 6px;
  padding-left: ${({ paddingLeft }) => `${paddingLeft}px`};
  cursor: pointer;
  ${({ foundation }) => foundation?.rounding?.round6}

  ${({ active }) => (active ? activeItemStyle : nonActiveItemStyle)}
  ${({ open }) => (open ? null : closedItemStyle)}

  ${({ interpolation }) => interpolation}
`
