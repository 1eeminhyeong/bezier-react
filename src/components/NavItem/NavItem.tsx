/* External dependencies */
import React, { useCallback, useMemo, memo } from 'react'
import { noop, isNil } from 'lodash-es'

/* Internal dependencies */
import useDepthContext, { DepthContext } from 'Hooks/useDepthContext'
import { IconSize, isIconName } from 'Components/Icon'
import type NavItemProps from './NavItem.types'
import {
  ItemWrapper,
  LeftIconWrapper,
  StyledIcon,
  ChevronWrapper,
  ContentWrapper,
  RightContentWrapper,
} from './NavItem.styled'

const DEFAULT_INDENT = 30
const DEFAULT_PADDING_LEFT = 6
export const NAV_ITEM_TEST_ID = 'bezier-react-nav-item'

function NavItem(
  {
    as,
    testId = NAV_ITEM_TEST_ID,
    name,
    style,
    className,
    interpolation,
    children,
    content,
    rightContent,
    leftIcon,
    leftIconColor = 'txt-black-dark',
    indent = DEFAULT_INDENT,
    active,
    open,
    onClick = noop,
  }: NavItemProps,
) {
  const depthInfo = useDepthContext((parent) => ({
    depth: parent.depth + 1,
    paddingLeft: (parent.paddingLeft ?? (DEFAULT_PADDING_LEFT - indent)) as number + indent,
  }))

  const handleClickItem = useCallback((e?: React.MouseEvent) => {
    onClick(e, name)
  }, [
    name,
    onClick,
  ])

  const leftComponent = useMemo(() => {
    if (!isNil(leftIcon) && isIconName(leftIcon)) {
      return (
        <LeftIconWrapper>
          <StyledIcon
            name={leftIcon}
            size={IconSize.S}
            color={leftIconColor}
          />
        </LeftIconWrapper>
      )
    }

    return null
  }, [
    leftIcon,
    leftIconColor,
  ])

  const contentComponent = useMemo(() => (
    <ContentWrapper>
      { content }
    </ContentWrapper>
  ), [content])

  const chevronComponent = useMemo(() => {
    const chevronIconName = open ? 'chevron-small-down' : 'chevron-small-right'

    return (
      <ChevronWrapper>
        <StyledIcon
          name={chevronIconName}
          size={IconSize.S}
          color="txt-black-dark"
        />
      </ChevronWrapper>
    )
  }, [open])

  const rightComponent = useMemo(() => (
    <RightContentWrapper>
      { rightContent }
    </RightContentWrapper>
  ), [rightContent])

  const showChevron = !isNil(children)

  return (
    <>
      <ItemWrapper
        as={as}
        active={active}
        open={open}
        style={style}
        className={className}
        interpolation={interpolation}
        paddingLeft={depthInfo.paddingLeft ?? 0}
        onClick={handleClickItem}
        data-testid={testId}
      >
        { leftComponent }
        { contentComponent }
        { showChevron && chevronComponent }
        { rightContent && rightComponent }
      </ItemWrapper>
      {
        open && (
          <DepthContext.Provider value={depthInfo}>
            { children }
          </DepthContext.Provider>
        )
      }
    </>
  )
}

export default memo(NavItem)
