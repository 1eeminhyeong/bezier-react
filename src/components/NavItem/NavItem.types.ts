/* Internal dependencies */
import type {
  BezierComponentProps,
  ChildrenProps,
  ContentProps,
  SideContentProps,
  ActivatableProps,
  AdditionalColorProps,
} from 'Types/ComponentProps'
import { IconName } from 'Components/Icon'

interface NavItemOptions {
  open?: boolean
  leftIcon?: IconName
  name?: string
  indent?: number
  onClick?: (e?: React.MouseEvent, name?: string) => void
}

export default interface NavItemProps extends
  BezierComponentProps,
  ChildrenProps,
  ContentProps,
  Pick<SideContentProps, 'rightContent'>,
  AdditionalColorProps<'leftIcon'>,
  Pick<ActivatableProps, 'active'>,
  NavItemOptions {}
