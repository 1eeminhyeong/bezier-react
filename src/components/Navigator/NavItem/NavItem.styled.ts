/* Internal dependencies */
import { styled, css, ellipsis } from 'Foundation'
import type { InterpolationProps } from 'Types/Foundation'
import { Text } from 'Components/Text'

interface WrapperProps extends InterpolationProps {
  active: boolean
}

export const LeftIconWrapper = styled.div`
  display: flex;
  align-items: center;
  width: 20px;
  margin-right: 8px;
`

export const ContentWrapper = styled(Text)`
  ${ellipsis()}
`

export const RightContentWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: auto;
`

const activeItemStyle = css`
  color: ${({ foundation }) => foundation?.theme?.['bgtxt-blue-normal']};
  background-color: ${({ foundation }) => foundation?.theme?.['bgtxt-blue-lightest']};
`

const nonActiveItemStyle = css`
  &:hover {
    background-color: ${({ foundation }) => foundation?.theme?.['bg-black-lighter']};
  }
`

export const ItemWrapper = styled.div<WrapperProps>`
  display: flex;
  align-items: center;
  height: 28px;
  padding: 0 6px;
  cursor: pointer;
  ${({ foundation }) => foundation?.rounding?.round6}

  ${({ active }) => (active ? activeItemStyle : nonActiveItemStyle)}

  ${({ interpolation }) => interpolation}
`
