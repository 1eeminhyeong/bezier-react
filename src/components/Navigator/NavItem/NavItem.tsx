/* External dependencies */
import React, { useCallback, memo } from 'react'
import { noop } from 'lodash-es'

/* Internal dependencies */
import { Typography } from 'Foundation'
import { Icon, IconSize, isIconName } from 'Components/Icon'
import type NavItemProps from './NavItem.types'
import {
  Item,
  Wrapper,
  ContentWrapper,
  RightContentWrapper,
  LeftIconWrapper,
} from './NavItem.styled'

export const NAV_ITEM_TEST_ID = 'bezier-react-nav-item'

function NavItem({
  as,
  testId = NAV_ITEM_TEST_ID,
  name,
  style,
  className,
  interpolation,
  content,
  href,
  target = '_self',
  rightContent,
  leftIcon,
  leftIconColor = 'txt-black-dark',
  active,
  onClick = noop,
}: NavItemProps) {
  const handleClickItem = useCallback((e?: React.MouseEvent) => {
    onClick(e, name)
  }, [
    name,
    onClick,
  ])

  const showLeftIcon = isIconName(leftIcon)

  return (
    <Wrapper
      role="menuitem"
    >
      <Item
        as={as}
        active={active}
        style={style}
        className={className}
        interpolation={interpolation}
        onClick={handleClickItem}
        data-testid={testId}
        href={href}
        target={target}
      >
        <LeftIconWrapper>
          { showLeftIcon && (
            <Icon
              name={leftIcon}
              size={IconSize.S}
              color={leftIconColor}
            />
          ) }
        </LeftIconWrapper>

        <ContentWrapper typo={Typography.Size14}>
          { content }
        </ContentWrapper>

        { rightContent && (
          <RightContentWrapper>
            { rightContent }
          </RightContentWrapper>
        ) }
      </Item>
    </Wrapper>
  )
}

export default memo(NavItem)
