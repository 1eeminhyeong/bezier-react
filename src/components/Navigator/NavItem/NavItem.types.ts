/* Internal dependencies */
import type {
  BezierComponentProps,
  ContentProps,
  LinkProps,
  SideContentProps,
  ActivatableProps,
  AdditionalColorProps,
} from 'Types/ComponentProps'
import { IconName } from 'Components/Icon'

interface NavItemOptions {
  leftIcon?: IconName
  name: string
  target?: string
  onClick?: (e?: React.MouseEvent, name?: string) => void
}

export default interface NavItemProps extends
  BezierComponentProps,
  ContentProps,
  LinkProps,
  Pick<SideContentProps, 'rightContent'>,
  AdditionalColorProps<'leftIcon'>,
  Pick<ActivatableProps, 'active'>,
  NavItemOptions {}
