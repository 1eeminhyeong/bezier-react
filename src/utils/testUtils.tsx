/* External dependencies */
import React, { ReactElement } from 'react'
import { render as baseRender, RenderOptions } from '@testing-library/react'

/* Internal dependencies */
import { FoundationProvider, LightFoundation } from 'Foundation'
import { ChildrenComponentProps } from 'Types/ComponentProps'

function TestProviders({ children }: ChildrenComponentProps) {
  return (
    <FoundationProvider foundation={LightFoundation}>
      { children }
    </FoundationProvider>
  )
}

export function render(
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) {
  return baseRender(ui, { wrapper: TestProviders, ...options })
}
