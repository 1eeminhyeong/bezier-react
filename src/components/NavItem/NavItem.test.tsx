/* External dependencies */
import React from 'react'

/* Internal dependencies */
import { render } from 'Utils/testUtils'
import NavItem, { NAV_ITEM_TEST_ID } from './NavItem'
import type NavItemProps from './NavItem.types'

describe('NavItem Test >', () => {
  let props: NavItemProps

  beforeEach(() => {
    props = {
      content: 'test-content',
    }
  })

  const renderNavItem = (optionProps?: NavItemProps) =>
    render(<NavItem {...props} {...optionProps} />)

  it('Snapshot > ', () => {
    const { getByTestId } = renderNavItem()

    const rendered = getByTestId(NAV_ITEM_TEST_ID)

    expect(rendered).toMatchSnapshot()
  })
})
