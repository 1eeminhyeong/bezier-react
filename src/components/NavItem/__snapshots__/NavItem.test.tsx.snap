// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItem Test > Snapshot >  1`] = `
.c1 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 28px;
  padding-right: 6px;
  padding-left: 6px;
  font-size: 14px;
  font-weight: 500;
  color: #000000D9;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  border-radius: 6px;
}

.c0:hover {
  background-color: #0000000D;
}

.c0 .sc-hKFxyN {
  visibility: hidden;
}

.c0:hover .sc-hKFxyN {
  visibility: visible;
}

<div
  class="c0"
  data-testid="bezier-react-nav-item"
>
  <div
    class="c1"
  >
    test-content
  </div>
</div>
`;
